da/kaspa/kaspa.go

```diff
@@ -0,0 +1,351 @@
+func (c *DataAvailabilityLayerClient) SubmitBatch(batch *types.Batch) da.ResultSubmitBatch {
+	blob, err := batch.MarshalBinary()
+	if err != nil {
+		return da.ResultSubmitBatch{
+			BaseResult: da.BaseResult{
+				Code:    da.StatusError,
+				Message: err.Error(),
+			},
+		}
+	}
+
+	c.logger.Debug("Submitting to da batch with size", "size", len(blob))
+	return c.submitBatchLoop(blob)
+}
+
+// submitBatchLoop tries submitting the batch, keep trying indefinitely.
+func (c *DataAvailabilityLayerClient) submitBatchLoop(dataBlob []byte) da.ResultSubmitBatch {
+	backoff := c.config.Backoff.Backoff()
+
+	for {
+		select {
+		case <-c.ctx.Done():
+			return da.ResultSubmitBatch{}
+		default:
+
+			// TODO: differentiate between unrecoverable and recoverable errors.(https://github.com/dymensionxyz/dymint/issues/1416)
+			txHash, blobHash, err := c.client.SubmitBlob(dataBlob)
+			if err != nil {
+				metrics.RollappConsecutiveFailedDASubmission.Inc()
+				c.logger.Error("submit blob", "err", err)
+				backoff.Sleep()
+				continue
+			}
+			metrics.RollappConsecutiveFailedDASubmission.Set(0)
+			submitMetadata := &SubmitMetaData{
+				TxHash:   txHash,
+				BlobHash: blobHash,
+			}
+
+			err = retry.Do(
+				func() error {
+					result := c.checkBatchAvailability(submitMetadata)
+					return result.Error
+				},
+				retry.Context(c.ctx),
+				retry.LastErrorOnly(true),
+				retry.Delay(c.batchRetryDelay),
+				retry.DelayType(retry.FixedDelay),
+				retry.Attempts(c.batchRetryAttempts),
+			)
+			if err != nil {
+				c.logger.Error("Check batch availability: submitted batch but did not get availability success status.", "error", err)
+				metrics.RollappConsecutiveFailedDASubmission.Inc()
+				backoff.Sleep()
+				continue
+			}
+
+			c.logger.Debug("Successfully submitted batch.")
+			return da.ResultSubmitBatch{
+				BaseResult: da.BaseResult{
+					Code:    da.StatusSuccess,
+					Message: "success",
+				},
+				SubmitMetaData: &da.DASubmitMetaData{
+					DAPath: submitMetadata.ToPath(),
+					Client: da.Kaspa,
+				},
+			}
+		}
+	}
+}
+
+// RetrieveBatches retrieves batch from DataAvailabilityLayerClient instance.
+func (c *DataAvailabilityLayerClient) RetrieveBatches(daPath string) da.ResultRetrieveBatch {
+	daMetaData := &SubmitMetaData{}
+	daMetaData, err := daMetaData.FromPath(daPath)
+	if err != nil {
+		return da.ResultRetrieveBatch{BaseResult: da.BaseResult{Code: da.StatusError, Message: "wrong da path", Error: err}}
+	}
+
+	blob, err := c.client.GetBlob(daMetaData.TxHash)
+	if err != nil {
+		return da.ResultRetrieveBatch{
+			BaseResult: da.BaseResult{
+				Code:    da.StatusError,
+				Message: "Blob not found",
+				Error:   err,
+			},
+		}
+	}
+
+	batch := &types.Batch{}
+	err = batch.UnmarshalBinary(blob)
+	if err != nil {
+		c.logger.Error("Unmarshaling blob", "error", err)
+		return da.ResultRetrieveBatch{
+			BaseResult: da.BaseResult{
+				Code:    da.StatusError,
+				Message: "Err Unmarshal",
+				Error:   err,
```

Comment:
isn't unmarshaling failure should return `ErrBlobNotParsed`? as we're checking this in sl validator.